### DISCUSSION TITLE
Building from source on Windows11 with conda

### DISCUSSION THREAD

**Discussion by DavidACosgrove** [rdkit/rdkit#6148] [Development]
I am having problems building the master branch on a Windows 11 machine via conda.  I have done it successfully in the past, but today the force is not with me.  
I've done:
```
 conda create -n rdkit_build2 python=3.9
 conda activate rdkit_build2
 conda install -y cmake cairo pillow eigen pkg-config
 conda install -y boost-cpp boost py-boost
 conda install -y swig=3.0
```
Then
```
cmake -DBoost_NO_BOOST_CMAKE=TRUE -DBOOST_ROOT="$CONDA_PREFIX" \
DRDK_BUILD_PYTHON_WRAPPERS=OFF -DRDK_USE_BOOST_REGEX=ON -DRDK_BUILD_SWIG_WRAPPERS=ON \
DRDK_BUILD_JAVA_WRAPPER=ON -DRDK_BUILD_CSHARP_WRAPPER=ON ..
```
works fine, and goes on to compile happily.
However, if I try and build the python wrappers:
```
cmake -DBoost_NO_BOOST_CMAKE=TRUE -DBOOST_ROOT="$CONDA_PREFIX" \
DRDK_BUILD_PYTHON_WRAPPERS=ON -DRDK_USE_BOOST_REGEX=ON -DRDK_BUILD_SWIG_WRAPPERS=ON \
DRDK_BUILD_JAVA_WRAPPER=ON -DRDK_BUILD_CSHARP_WRAPPER=ON ..
```
I get
```
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.22621.
-- The C compiler identification is MSVC 19.34.31937.0
-- The CXX compiler identification is MSVC 19.34.31937.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
Disabling boost::stacktrace on non-linux platform
-- Found Catch2 source in C:/Users/david/my_rdkit/External/catch/catch
CATCH: C:/Users/david/my_rdkit/External/catch/catch/single_include/catch2
-- Found PythonInterp: C:/Users/david/anaconda3/envs/rdkit_build2/python.exe (found version "3.9.16")
-- Found PythonLibs: C:/Users/david/anaconda3/envs/rdkit_build2/libs/python39.lib (found version "3.9.16")
CMake Warning at C:/Users/david/anaconda3/envs/rdkit_build2/Library/share/cmake-3.22/Modules/FindBoost.cmake:2201 (message):
  No header defined for python-py39; skipping header check (note: header-only
  libraries have no designated component)
Call Stack (most recent call first):
  CMakeLists.txt:305 (find_package)


CMake Error at C:/Users/david/anaconda3/envs/rdkit_build2/Library/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Boost (missing: python) (found suitable version "1.73.0",
  minimum required is "1.58.0")
Call Stack (most recent call first):
  C:/Users/david/anaconda3/envs/rdkit_build2/Library/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  C:/Users/david/anaconda3/envs/rdkit_build2/Library/share/cmake-3.22/Modules/FindBoost.cmake:2360 (find_package_handle_standard_args)
  CMakeLists.txt:313 (find_package)


-- Configuring incomplete, errors occurred!
See also "C:/Users/david/my_rdkit/csharp_build/CMakeFiles/CMakeOutput.log".
See also "C:/Users/david/my_rdkit/csharp_build/CMakeFiles/CMakeError.log".
```

This sort of incantation has always worked fine for me on Linux and macOS.  Can anyone point out the error, or supply a CMake command that works for them?  

Thanks,
Dave

**Comment by DavidACosgrove** [comment]
I seem to have had success.  I'm posting the successful commands in case it's useful for other people.  It has taken a lot of frustrating experimentation to get this far!  It may have unnecessary steps, but it does at least build the Python wrappers.

I am using the git bash shell.
I made a fresh installation of mamba and installed the necessary development packages:
```
mamba create -n rdkit.build python=3.9
mamba activate rdkit.build
mamba install -y numpy matplotlib
mamba install -y cmake cairo pillow eigen pkg-config
mamba install -y boost-cpp boost swig
```
Then make and cd to the build directory and run cmake to configure and build:
```
mkdir build
cd build
cmake -DPy_ENABLE_SHARED=1 \
-DBOOST_ROOT="$CONDA_PREFIX" \
-DBoost_NO_SYSTEM_PATHS=ON \
-DBoost_NO_BOOST_CMAKE=TRUE \
-DRDK_INSTALL_INTREE=ON \
-DRDK_INSTALL_STATIC_LIBS=OFF \
-DRDK_BUILD_CPP_TESTS=ON \
-DRDK_BUILD_CAIRO_SUPPORT=ON \
-DRDK_BUILD_FREETYPE_SUPPORT=ON \
-DBOOST_ROOT=$CONDA_PREFIX \
-DRDK_BOOST_PYTHON3_NAME=python39 \
-DBoost_INCLUDE_DIR=$CONDA_PREFIX/Library/include \
..
cmake --build . --config Release --target install -j 10
```
It all built and installed - I have a whole load of .pyd files under $RDBASE.
Then I added the relevant directories to PATH and PYTHONPATH:
```
export PATH=/c/Users/david/rdkit/rdkit:${PATH}
export RDBASE=/c/Users/david/rdkit
export PYTHONPATH=$RDBASE
```
And it all seems to work:
```
$ python -c "from rdkit import rdBase ; print(rdBase.rdkitVersion)"
2023.03.1pre
```

**Reply by kienerj** [reply]
Hi David, 

I'm trying to build rdkit (and cartridge) on windows and followed your procedure above. the build seemed to have completed successfully.  However your simple test "script" fails with the dreaded 

    ImportError: DLL load failed while importing rdBase: The specified module could not be found.

I'm pretty sure PATH should be fine. especially I can see the `$CONDA_PREFIX\Library\bin` which contains boost dlls in the path. I'm not sure which dll is causing the issue really. I checked with  `DependenciesGui.exe` and only things it says missing are boost related and python dll which all are in the PATH. 

Therefore some input on how to figure out what the missing dll is would be very helpful.

**Reply by ptosco** [reply]
@kienerj Hi Joos, if you have built support for `cairo` you will also need `libcairo-2.dll` to be in your `PATH` - it will not show up in `DependenciesGui.exe` as it not linked, but it needs to be available at runtime and cause that error message if not found.
On my Windows machine I have `libcairo-2.dll` in my MSYS64 installation and add `C:\MSYS64\mingw64\bin` to `PATH` for that reason before using RDKit.
Unfortunately there is no way to persuade Windows to tell you which the missing DLL is. When you are desperate, [ProcessMonitor](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) is a great tool to figure that out (requires admin privileges to be run) - if you keep it open while you attempt to import `rdBase` you will see which DLL it is looking for without finding it.

**Reply by kienerj** [reply]
Hi Paolo, 

I got hold of libcairo-2.dll and put it in the path but that did not solve the issue.

With ProcessMonitor I figured out for whatever reason the rdkit dlls seem to have to be at certain predefined locations. Adding their location `rdkit\lib` or `rdkit\build\bin\Release` doesn't help. These paths are not searched:

![image](https://user-images.githubusercontent.com/1567309/228136424-cb7df7ef-1976-467d-8918-9b119751d9e7.png)

Copy&pasting them into one of the listed directories solves the issue, eg I get a rdkit version number and no more DLL error.

Still I'm unclear why folders listed in PATH are ignored? I can also add them at the start or end of the PATH. It doesn't matter.

**Reply by ptosco** [reply]
@kienerj Does calling `os.add_dll_directory(your_DLL_path)` prior to importing `rdBase` fix the problem?
If so, this would be due to a [change introduced in Python 3.8](https://docs.python.org/3/library/os.html#os.add_dll_directory).
Later tonight I will try and figure out why I don't incur into the same problem on my home Windows build machine, where I am definitely building RDKit against Python >3.8, and I am loading DLLs from a custom directory that I add to `PATH`.

**Reply by kienerj** [reply]
Indeed if I use `os.add_dll_directory(r"C:\Users\kienerj\rdkit\lib")` before `from rdkit import rdBase` it works. With or without that directory in the PATH.

And the python documentation says this for anyone else reading this discussion:

> DLL dependencies for extension modules and DLLs loaded with [ctypes](https://docs.python.org/3/library/ctypes.html#module-ctypes) on Windows are now resolved more securely. Only the system paths, the directory containing the DLL or PYD file, and directories added with [add_dll_directory()](https://docs.python.org/3/library/os.html#os.add_dll_directory) are searched for load-time dependencies. **Specifically, PATH and the current working directory are no longer used, and modifications to these will no longer have any effect on normal DLL resolution**.

It also explains why the cartridge just worked as that doesn't use python and hence uses PATH as expected.

**Comment by DavidACosgrove** [comment]
Hi @kienerj,
These notes comprise pretty much the sum total of my knowledge of building RDKit on a Windows machine, I'm afraid.  The missing DLL is probably rdBase.pyd, so you should make sure that that is in PATH.  In my build, that is in $RDBASE/rdkit. 
 Other than that, I don't think I'll be able to help very much.
Dave

**Reply by kienerj** [reply]
Ok thanks. That is already in the PATH. 

Also interestingly enough building and running the cartridge via conda seems to have worked:

![image](https://user-images.githubusercontent.com/1567309/227943808-2d6dcf08-a153-4918-9c03-9f7b82c0727f.png)

Eg starting postgresql from the exact same command prompt as running the python snippet works while the python snippet does not work. I guess I'm waiting for Gregs blog post about building on Windows then.
